### PR TITLE
net/ipfwd: Support ICMP/ICMPv6 error reply when forwarding

### PR DESCRIPTION
**Pull request by wengzhe**
## Summary

When forwarding packets, reply ICMP for error `-ENETUNREACH`, `-EFBIG` and `-EMULTIHOP`.

patches included:
- net/ipv4_input: Set IPv4 flag at the same place as ipv6_input
- net/ipfwd: Support ICMP error reply when forwarding IPv4
- net/ipfwd: Support ICMPv6 error reply when forwarding IPv6
- net/icmpv6: Fix icmpv6_reply function

## Impact

When forwarding packets, NuttX will reply ICMP for some errors now.

## Testing
Tested on sim enabled 2 tap devices with mtu of 1500 and 300, by `ping -t1` and `ping -s 500`

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by anchao — APPROVED** [review]

File: net/icmpv6/icmpv6_reply.c [hunk 1]
```diff
@@ -112,15 +112,19 @@ void icmpv6_reply(FAR struct net_driver_s *dev, int type, int code, int data)
 
   dev->d_len = ipicmplen + datalen;
 
+  /* Copy fields from original packet */
+
+  memmove(icmpv6 + 1, ipv6, datalen);
```
Comment (acassis) [inline net/icmpv6/icmpv6_reply.c:117]: Why to skip a byte in the icmpv6 destination?

Comment (wengzhe) [reply net/icmpv6/icmpv6_reply.c:117]: It's skipping the whole header, to copy origin packet as the payload of icmpv6.
It's removed by accident and I'm just restoring the code from [previous version](https://github.com/apache/nuttx/blob/releases/11.0/net/icmpv6/icmpv6_reply.c#L117-L119).

File: net/ipforward/ipv4_forward.c [hunk 2]
```diff
@@ -505,6 +502,37 @@ int ipv4_forward(FAR struct net_driver_s *dev, FAR struct ipv4_hdr_s *ipv4)
 
 drop:
   ipv4_dropstats(ipv4);
+
+#ifdef CONFIG_NET_ICMP
+#ifdef CONFIG_NET_NAT
```
[the remark this answers is not in the document]
Comment (wengzhe) [reply net/ipforward/ipv4_forward.c]: Done.

File: net/ipforward/ipv4_forward.c [hunk 3]
```diff
@@ -505,6 +502,37 @@ int ipv4_forward(FAR struct net_driver_s *dev, FAR struct ipv4_hdr_s *ipv4)
...
+  /* Before we reply ICMP, call NAT outbound to try to translate destination
+   * address & port back to original status.
+   */
+
+  if (ret == -ENETUNREACH || ret == -EFBIG || ret == -EMULTIHOP)
+    {
+      ipv4_nat_outbound(dev, ipv4, NAT_MANIP_DST);
+    }
+#endif /* CONFIG_NET_NAT */
```
[the remark this answers is not in the document]
Comment (wengzhe) [reply net/ipforward/ipv4_forward.c]: Done.

File: net/ipforward/ipv4_forward.c [hunk 4]
```diff
@@ -505,6 +502,37 @@ int ipv4_forward(FAR struct net_driver_s *dev, FAR struct ipv4_hdr_s *ipv4)
...
+        return OK;
+
+      case -EFBIG:
+        icmp_reply(dev, ICMP_DEST_UNREACHABLE, ICMP_FRAG_NEEDED);
+        return OK;
+
+      case -EMULTIHOP:
+        icmp_reply(dev, ICMP_TIME_EXCEEDED, 0);
+        return OK;
+    }
```
[the remark this answers is not in the document]
Comment (wengzhe) [reply net/ipforward/ipv4_forward.c:536]: Done.

[the remark this answers is not in the document]
Comment (wengzhe) [reply net/ipforward/ipv4_forward.c]: Yeah it's better on case matching (without another `if` with probability of mismatch), but the params of `icmp_reply` is really too long (much more difficult if we want to add more cases), I'm still considering how to improve this.

[the remark this answers is not in the document]
Comment (wengzhe) [reply net/ipforward/ipv4_forward.c]: Tried to add another label `reply` for better maintainability, maybe.

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**